draft-polli-ratelimit-headers.md

```diff
@@ -407,6 +407,37 @@ with the lower `RateLimit-Remaining` values.
 Under certain conditions, a server MAY artificially lower `RateLimit` field values between subsequent requests,
 eg. to respond to Denial of Service attacks or in case of resource saturation.
 
+# Intermediaries
+
+An intermediary MUST NOT alter the RateLimit headers values
+in such a way as to communicate a more permissive quota-policy;
+this includes removing the RateLimit headers.
+
+An intermediary MAY alter the RateLimit headers
+in such a way as to communicate a more restrictive quota-policy when:
+
+- it is aware of the quota-unit semantic used by the Origin Server;
+- it implements this specification and enforces a quota-policy which
+  is more restrictive than the one conveyed in the headers.
+
+An intermediary that does not implement this specification
+SHOULD forward a request even if it presumes that that request
+might not be serviced;
+the server returning the RateLimit headers is the sole responsible
+of enforcing the communicated quota-policy.
+
+This specification does not mandate any behavior on intermediaries
+respect to retries,
+nor requires that intermediaries have any role in respecting quota-policies.
+For example, it is legitimate for a proxy to retransmit a request
+without notifying the client, and thus consuming quota-units.
+
+# Caching
+
+The RateLimit headers MUST NOT be cached because their values are ephemeral.
```

Comment:
These are going to get cached.  We probably need some other way to identify stale values.

Reply:
Ok, suggestions welcome!

Is the intermediary part ok?

Reply:
I don't see how to proceed without adding some smarts to caches about these limits. If a cache is going to keep responding with stale data, then these headers are likely to matter a lot more to it than to the client originating the requests.

Reply:
@unleashed we should find a solution that avoids modifying caches.

If a user-agent uses cached ratelimit headers, at worst it will be throttled out (which is the same behavior that you'll get without the headers). In all the other cases, it will land on the cache.

If the user-agent feels it appropriate, it can use `Date` and `RateLimit-Reset` to check if the reset moment has passed or not.

Reply:
@unleashed PTAL

Reply:
Yes, I was thinking about a way to relate the headers to a specific moment in time to let the user agent figure that the information is stale.

Given that having intermediaries deal with these headers is unlikely to happen anytime soon and the fact caches are going to keep passing the headers around, we could just add a provision to disregard the headers in case the user agent notices the information is stale.

The problem I see with the concept of stale information is how to determine it is actually stale. Using the `RateLimit-Reset` header has the problem that its value is relative and we allow an origin to return completely different data the next time a request is performed, even if it is within the described time window, so in some pathological cases a user agent might have trouble deducing whether these headers are stale or not.

Reply:
```suggestion
The RateLimit headers MAY be ignored by caches because their values are ephemeral.
```

> how to determine it is actually stale

imho `Date` is the key. Both with a relative or an absolute timestamp, it's only via `Date` (or other response-timestamp field) that you can determine when the value was generated.

I agree that in patological use-cases clients should rely on the `Limit` and `window`, but this still happens with `retry-after`.

Reply:
The change suggested looks fine. However I can see implementations trying to find out an algorithm to identify stale data once they become aware of the role intermediaries can play. Should we try to prescribe or hint at solutions in the text?

Reply:
We can provide guidance and ask the httpapi wg if those parts are needed or not. Probably warning implementors and let them decide is a reasonable way to do it.

Reply:
Avoid normative statements

```suggestion
```